test/cpp/jit/test_gpu_shift.cpp

```diff
@@ -1907,6 +1907,297 @@ TEST(NVFuserTest, FusionShiftBcast3_CUDA) {
 }
 
+// See issue #893
+TEST(NVFuserTest, FusionShiftSyncPlacement1_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+  auto tv1 = add(tv0, new Double(1));
+  auto tv2 = add(tv0, new Double(2));
+  auto tv3 = add(tv1, tv2);
+  auto tv4 = shift(tv3, {0, 1});
+  fusion.addOutput(tv4);
+
+  tv4->split(1, 8);
+  tv0->computeAt(tv4, 2);
+
+  tv2->computeAt(tv3, -1);
+
+  tv1->setMemoryType(MemoryType::Shared);
+  tv3->setMemoryType(MemoryType::Shared);
+
+  tv1->axis(-1)->parallelize(ParallelType::TIDx);
+  tv3->axis(-1)->parallelize(ParallelType::TIDx);
+  tv4->axis(-1)->parallelize(ParallelType::TIDx);
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+
+  int numel_x = 99;
+  int numel_y = 101;
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({numel_x, numel_y}, options);
+  std::vector<IValue> inputs = {t0};
+  auto outputs = fe.runFusion(inputs);
+
+  auto t1 = t0 + 1;
+  auto t2 = t0 + 2;
+  auto t3 = add(t1, t2);
+  auto t4 = shift(t3, {0, 1});
+
+  testValidate(&fusion, outputs, inputs, {t4}, __LINE__, __FILE__);
+}
+
+// See issue #893. Top-level placement.
+TEST(NVFuserTest, FusionShiftSyncPlacement2_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(1);
+  fusion.addInput(tv0);
+  auto tv1 = add(tv0, new Double(1));
+  auto tv2 = add(tv0, new Double(2));
+  auto tv3 = add(tv1, tv2);
+  auto tv4 = shift(tv3, {1});
+  fusion.addOutput(tv4);
+
+  tv2->computeAt(tv3, -1);
+
+  tv1->setMemoryType(MemoryType::Shared);
+  tv3->setMemoryType(MemoryType::Shared);
+
+  tv1->axis(-1)->parallelize(ParallelType::TIDx);
+  tv3->axis(-1)->parallelize(ParallelType::TIDx);
+  tv4->axis(-1)->parallelize(ParallelType::TIDx);
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+
+  int numel_x = 99;
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({numel_x}, options);
+  std::vector<IValue> inputs = {t0};
+  auto outputs = fe.runFusion(inputs);
+
+  auto t1 = t0 + 1;
+  auto t2 = t0 + 2;
+  auto t3 = add(t1, t2);
+  auto t4 = shift(t3, {1});
+
+  testValidate(&fusion, outputs, inputs, {t4}, __LINE__, __FILE__);
+}
+
+TEST(NVFuserTest, FusionShiftSyncPlacement3_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(1);
+  fusion.addInput(tv0);
+  auto tv1 = add(tv0, new Double(1));
+  auto tv2 = add(tv1, new Double(2));
+  auto tv3 = shift(tv2, {1});
+  fusion.addOutput(tv3);
+
+  // This doesn't work. syncthreads is needed between tv1 and tv2, but
+  // both the loop extent of both tv1 and tv2 has halo, so the loop is
+  // not eliminated even though it is parallelized. Moving syncthreads
+  // out of the loop would make it placed before tv1, which would make
+  // it meaningless.
+  // Ideally, an exception should be through at this computeAt, but at
```

Comment:
Technically, that's true, but the current sync insertion logic does not differentiate whether there really exists dependencies between threads through a smem buffer. This can be fixed.

Reply:
Thanks, yeah that makes sense. Was just asking since it's being placed as a test may be worth mentioning that if we change the behavior. I think it would be good to reduce syncthreads when not needed, but that can definitely be left for later work.